File: Metal/FrontEnd/Parse/Grammar/Utilities/ASTPrinter.cs
﻿using System;
using System.Text;
using Metal.FrontEnd.Parse.Grammar;

namespace Metal {
  class ASTPrinter : Expression.IVisitor<string> {
    public string Print(Expression expression) {
      return expression.Accept(this);
    }

    public string Visit(Expression.Literal expression) {
      return expression.Value.ToString();
    }

    public string Visit(Expression.Unary expression) {
      return Parenthesize(expression.Operator.Lexeme, expression.Right);
    }

    public string Visit(Expression.Binary expression) {
      return Parenthesize(expression.Operator.Lexeme, expression.Left, expression.Right);
    }

    public string Visit(Expression.Parenthesized expression) {
      return Parenthesize("parenthesized", expression.Center);
    }

    public string Visit(Expression.Assign expression) {
      throw new NotImplementedException();
    }

    public string Visit(Expression.Variable expression) {
      throw new NotImplementedException();
    }

    public string Visit(Expression.Logical expression) {
      throw new NotImplementedException();
    }

    private String Parenthesize(String name, params Expression[] expressions) {
      StringBuilder builder = new System.Text.StringBuilder();

      builder.Append("(").Append(name);
      foreach (var expression in expressions) {
        builder.Append(" ");
        builder.Append(expression.Accept(this));
      }
      builder.Append(")");

      return builder.ToString();
    }
  }
}

File: src/metal/utilities/ASTPrinter.java
package metal.utilities;

import metal.frontend.scanner.Token;
import metal.frontend.scanner.TokenType;
import metal.frontend.parser.grammar.Expression;
// Creates an unambiguous, if ugly, string representation of AST nodes.
public class ASTPrinter implements Expression.Visitor<String> {                     
  public String print(Expression expression) {                                            
    return expression.accept(this);                                          
  } 
  @Override                                                          
  public String visitBinaryExpression(Expression.Binary expression) {                  
    return parenthesize(expression.operator.lexeme, expression.left, expression.right);
  }

  @Override                                                          
  public String visitGroupExpression(Expression.Group expression) {              
    return parenthesize("group", expression.expression);                   
  }                                                                  

  @Override                                                          
  public String visitLiteralExpression(Expression.Literal expression) {                
    if (expression.value == null) return "nil";                            
    return expression.value.toString();                                    
  }                                                                  

  @Override                                                          
  public String visitUnaryExpression(Expression.Unary expression) {                    
    return parenthesize(expression.operator.lexeme, expression.right);           
  } 
  
  private String parenthesize(String name, Expression... expressions) {
    StringBuilder builder = new StringBuilder();

    builder.append("(").append(name);                      
    for (Expression expression : expressions) {                              
      builder.append(" ");                                 
      builder.append(expression.accept(this));                   
    }                                                      
    builder.append(")");                                   

    return builder.toString();                             
  } 

  public static void main(String[] args) {                 
    Expression expression = new Expression.Binary(                     
        new Expression.Unary(                                    
            new Token(TokenType.Operator, "-", null, 1, 2),      
            new Expression.Literal(123)),                        
        new Token(TokenType.Operator, "*", null, 1, 2),           
        new Expression.Group(                                 
            new Expression.Literal(45.67)));

    System.out.println(new ASTPrinter().print(expression));
  } 
}                                                                    